File: c++/deserialize.cpp
/*
	Deserialize - load bytecode produced by compiler.

	Copyright (c) 2022 Frank McIngvale, see LICENSE
*/
#include "deserialize.hpp"
#include "errors.hpp"

using namespace std;

static void string_replace(string &s, const char *from, const char *to) {
	size_t i;
	while((i = s.find(from)) != string::npos)
		s.replace(i, strlen(from), to);
}

// FYI -- stream from compiler will be a single list containing word definitions

// see compiler.verb:serialize-object, but briefly, I only have to deserialize the
// format that the compiler produces, not ANY arbitrary verbii object

// deserialize & return next object from stream
Object deserialize_stream(Interpreter *intr, ifstream &fileIn) {
	string line;
	if(getline(fileIn, line)) {
		switch(line[0]) {
			case 'i': return parseInt(line.substr(2));
			case 'f': return parseFloat(line.substr(2));
			case 'b': return parseBool(line.substr(2));
			case 'n': return newNull();
			case 'o': return newOpcode(parseInt(line.substr(2)).asInt());
			case 's':
				string_replace(line, "%32", " ");
				string_replace(line, "%09", "\t");
				string_replace(line, "%10", "\n");
				string_replace(line, "%13", "\r");
				string_replace(line, "%37", "%");
				return newString(line.substr(2));
			case 'y': return newSymbol(line.substr(2));
			case 'L': // list
				{
					int nr = parseInt(line.substr(2)).asInt();
					Object list = newList();
					for(int i=0; i<nr; ++i)
						list.data.objlist->push_back(deserialize_stream(intr,fileIn));
					return list;
				}
			case 'F': // lambda
				{
					Object list = deserialize_stream(intr,fileIn);
					if(!list.isList())
						throw LangError("Expecting list after F but got:" + list.fmtStackPrint());
					return newLambda(list.data.objlist);
				}
			case 'W': // word definition
				{
					string name = line.substr(2);
					Object list = deserialize_stream(intr,fileIn);
					if(!list.isList())
						throw LangError("Expecting list after W but got:" + list.fmtStackPrint());

					// insert into interpreter
					//cout << "Inserting word:>>>" << name << "<<<" << endl;
					//intr->WORDS[name.c_str()] = list.data.objlist;

					// do not allow overwriting words when deserializing
					intr->defineWord(name.c_str(), list.data.objlist, false);
					// this produces nothing extra
					return newVoid();
				}

			default:
				// this is a fatal error, so just print & exit -- trying to do an exception
				// makes the output weird
				//throw LangError("Unrecogized line in deserialize:" + line);
				printf("Unrecogized line in deserialize: %s\n", line.c_str());
				exit(1);
		}
	}
	return newVoid();
}
	

File: lua/deserialize.lua
--[[
	Deserialize - load bytecode from compiler and put into Interpreter.
	
	Copyright (c) 2022 Frank McIngvale, see LICENSE

	Ported from the Python implementation.
]]--

require("interpreter")

function trimCRLF(text)
	while text:sub(#text) == "\n" or text:sub(#text) == "\r" do
		text = text:sub(1,#text-1)
	end
	
	return text
end

function deserialize_stream(intr, fileIn)
	local line = fileIn:read('l')
	if line == fail then
		return new_Void()
	end
	-- get rid of extra CR/LF -- could have been written by another platform so the
	-- above :read('l') didn't remove the entire line ending
	line = trimCRLF(line)
	--print("LINE: " .. line)
	if line:sub(1,1) == "M" then
		-- ignore metadata and return NEXT object 
		return deserialize_stream(intr, fileIn)
	elseif line:sub(1,1) == "i" then
		return tonumber(line:sub(3))
	elseif line:sub(1,1) == "f" then
		return new_Float(tonumber(line:sub(3)))
	elseif line:sub(1,1) == "b" then
		return parseBool(line:sub(3))
	elseif line:sub(1,1) == "n" then
		return new_Null()
	elseif line:sub(1,1) == "o" then
		return opcode_from_packed(tonumber(line:sub(3)))
	elseif line:sub(1,1) == "s" then
		local s = line:sub(3)
		-- unescape string
		s = s:gsub("%%32", " ")
		s = s:gsub("%%09", "\t")
		s = s:gsub("%%10", "\n")
		s = s:gsub("%%13", "\r")
		s = s:gsub("%%37", "%%")
		return new_String(s)
	elseif line:sub(1,1) == "y" then
		return line:sub(3) -- symbols don't need escaping
	elseif line:sub(1,1) == "L" then
		local nr = tonumber(line:sub(3))
		local objs = {}
		for i=1,nr do
			objs[i] = deserialize_stream(intr,fileIn)
		end
		return objs
	elseif line:sub(1,1) == "F" then --lambda
		local objs = deserialize_stream(intr, fileIn)
		if isList(objs) then
			return new_Lambda(objs)
		else
			error(">>>Expecting list after 'F' but got: " .. fmtStackPrint(objs))
		end
	elseif line:sub(1,1) == "W" then
		local name = line:sub(3)
		local objs = deserialize_stream(intr, fileIn)
		if isList(objs) then
			intr:defineWord(name, objs, false) -- do not allow overwriting existing words
		else
			error(">>>Expecting list after 'W' but got: " .. fmtStackPrint(objs))
		end
	else
		error(">>>Unrecognized line while deserializing: " .. line)
	end
end


		
	
		
		
		
	
		
